Project 9

Samjamhas10 opened the pull request:
Project 9: Spots 

Added API integration to the Spots project:

- Connected front-end to server API
- Added ability to add/delete cards
- implemented user authorization
- Added like/unlike functionality
- Added profile edit and avatar update features 
- Added loading states "Saving..., Deleting..."
- All forms are working with proper validation

Ready for review

gennady-bars commented on src/utils/Api.js:
```diff
@@ -0,0 +1,111 @@
+  getInitialCards() {
+    return fetch(`${this._baseUrl}/cards`, {
+      headers: this._headers,
+    }).then((res) => {
+      if (res.ok) {
+        return res.json();
+      }
+      return Promise.reject(`Error: ${res.status}`);
```
```js
.then(res => {
                if (res.ok) {
                    return res.json();
                }
                return Promise.reject(`Error ${res.status}`);
            })

```

The code of `ok` checking is repeated in **every** request of the project. To avoid it, you need to make a special `_checkResponse` method that will check any response from the server if it’s ok or not.


```js
_checkResponse(res) {
        // here is the code of the checking
    }

```

And now you can replace redundant repeating with the only line:


```js
 .then(this._checkResponse)

```

Please, pay attention that you need to pass only the **reference** for the method rather than the call. 

**COULD BE IMPROVED**

You can make a special method for fetching and checking responses **not to duplicate** it in every request:

```js

  _request(url, options) {
    return fetch(url, options).then(this._checkResponse)
  }

```

And now you can safely replace `fetch` with `this._request` and remove repeated checking. And you don’t need to change anything else.

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,333 @@
+  },
+});
+
+api
+  .getAppInfo()
+  .then(([userInfo, cards]) => {
+    cards.forEach((item) => {
```
Good job that you get the whole info with `Promise.all`

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,333 @@
+
+// Edit form elements
+const editModal = document.querySelector("#edit-modal");
+const editFormElement = editModal.querySelector(".modal__form");
```
**COULD BE IMPROVED**

To avoid wasting resources on searching the forms in the DOM, you can efficiently access any **form** from `document.forms`, using their **unique** `id` attributes.

If a form has an `id` attribute of `profile-form` for example, instead of using
 
```js
const profileForm = document.querySelector("#profile-form");
```

you could instead use:

```js

const profileForm = document.forms["profile-form"];

```

This is not required, but it's good to know about another way of finding forms in the DOM.

More info here https://developer.mozilla.org/en-US/docs/Web/API/Document/forms

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,333 @@
+
+  const cardLikeBtn = cardElement.querySelector(".card__like-btn");
+  const deleteButton = cardElement.querySelector(".card__delete-btn");
+
```
* When I reload the page, all **red** hearts become **white** again. You need to check `isLiked`

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,333 @@
+    });
+}
+
+function handleEditFormSubmit(evt) {
+  evt.preventDefault();
+  const modal__submit = evt.submitter;
+  setButtonText({
+    btn: modal__submit,
+    isLoading: true,
+    defaultText: "Save",
+    loadingText: "Saving...",
```
**COULD BE IMPROVED**

If it’s interesting for you, here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **resetting** and **catching errors**


```js
// define a function for changing the button text. It accepts 4 params (the 2 last are optional with default texts)
export function renderLoading(isLoading, button, buttonText='Save', loadingText='Saving...') {
  if (isLoading) {
    button.textContent = loadingText
  } else {
    button.textContent = buttonText
  }
}

// define a universal function that accepts a request function, event and a default loading text 
function handleSubmit(request, evt, loadingText = 'Saving...') {
 // You need to prevent the default action in any submit handler
  evt.preventDefault();

  // the button is always available inside `event` as `submitter`
  const submitButton = evt.submitter;
  // fix the initial button text
  const initialText = submitButton.textContent;
  // change the button text before requesting
  renderLoading(true, submitButton, initialText, loadingText);
  // call the request function to be able to use the promise chain
  request()
    .then(() => {
      // any form should be reset after a successful response 
      // evt.target is the form in any submit handler		
      evt.target.reset();
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
    .catch(console.error)      

    // and in finally we need to stop loading
    .finally(() => {
      renderLoading(false, submitButton, initialText);
    });
}

```

Here is an example of handling the profile form submit:

```js
function handleProfileFormSubmit(evt) {
  // create a request function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally`
    return editProfile(nameInput.value, jobInput.value).then((userData) => {
      userName.textContent = userData.name;
      userJob.textContent = userData.about;
    });
  }
  // here we call handleSubmit passing the request and event (if you want a different loading text then you need to pass the 3rd param)
  handleSubmit(makeRequest, evt);
}

```

So, this way you can remove a lot of code duplicating. You will not need to search buttons, pass initial button texts and so on.

`handleSubmit` and `renderLoading` should be placed in `utils.js`, because they are utility functions

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,333 @@
+    .editAvatarInfo(avatarInput.value)
+    .then((data) => {
+      profileAvatar.src = data.avatar;
+      avatarFormElement.reset();
```
Please, disable the `avatar` submit button as well after submission


```js
      disableButton(evt.submitter, settings);

```

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,333 @@
+function handleAddCardSubmit(evt) {
+  evt.preventDefault();
+  const inputValues = { name: cardNameInput.value, link: cardLinkInput.value };
+  const modal__submit = evt.submitter;
```
`modal__submit`

Every variable in the project should be in `camelCase` 


`submitButton`

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,333 @@
+
+function handleEditFormSubmit(evt) {
+  evt.preventDefault();
+  const modal__submit = evt.submitter;
```
`modal__submit`

Every variable in the project should be in `camelCase` 


`submitButton`

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,333 @@
+
+function handleAvatarSubmit(evt) {
+  evt.preventDefault();
+  const modal__submit = evt.submitter;
```
`modal__submit`

Every variable in the project should be in `camelCase` 


`submitButton`

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,333 @@
+const avatarModalCloseBtn = avatarModal.querySelector(".modal__close-btn");
+const avatarInput = avatarModal.querySelector("#profile-avatar-input");
+
+modalCloseTypePreview.addEventListener("click", () => {
```
Please, delete the redundant `close icon` listeners because you handle them in line 300 **universally**

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,333 @@
+});
+
+// click the close button on the edit profile modal to close that modal
+editModalCloseBtn.addEventListener("click", () => {
```
Please, delete the redundant `close icon` listeners because you handle them in line 300 **universally**

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,333 @@
+avatarFormElement.addEventListener("submit", handleAvatarSubmit);
+deleteForm.addEventListener("submit", handleDeleteSubmit);
+
+cardModalCloseBtn.addEventListener("click", () => {
```
Please, delete the redundant `close icon` listeners because you handle them in line 300 **universally**

gennady-bars commented on src/pages/index.js:
```diff
@@ -0,0 +1,333 @@
+  closeModal(cardModal);
+});
+
+avatarModalCloseBtn.addEventListener("click", () => {
```
Please, delete the redundant `close icon` listeners because you handle them in line 300 **universally**

gennady-bars reviewed (commented):
**↓Expand to see the whole summary↓**

Hello, SamSam!

Great job on submitting your project 9! This time you’ve implemented `Api` integration. So you can create your own simple sites from now on! 

Please, scroll down the page on **github** to see my remarks right in the code

**Parts I really liked in your project:**

* Good job that you **close** modals only in `.then` blocks
* Good job that you **catch** possible errors at the end of server requests.
* Good job that you return the button text in `.finally`
* Good job that you get the whole info with `Promise.all`

There are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix.

**Needs correcting:**

* All comments in the code and the comments below:
* When I reload the page, all **red** hearts become **white** again. You need to check `isLiked`
* The code of `ok` checking is repeated in **every** request of the project. To avoid it, you need to make a special `_checkResponse` method that will check any response from the server if it’s ok or not.
* Please, disable the `avatar` submit button as well after submission
* Every variable in the project should be in `camelCase` 
* Please, delete the redundant `close icon` listeners because you handle them in line 300 **universally**



There are some comments in your project that can help you make it even better. Please check the `Could be improved` comments.

**A reminder:** It’s necessary to fix all comments to get the project accepted. And `Could be improved` comments are optional, they will be up to you if you want to implement them.

Your hard work on this project is truly commendable. While there are still some adjustments needed. If you need any assistance with the feedback or understanding certain concepts, our tutors are here to help. Keep moving forward and maintain your dedication — you are doing an amazing job!✨

Samjamhas10 commented:
I've updated the code based on the review feedback: 

- Fixed the like button functionality
- Added error handling for the API request

Ready for review

gennady-bars reviewed (approved):
You’ve done a great job!

Your project has been accepted. 

Good luck!
